### PR TITLE
[jit] Document unsupported types

### PR DESCRIPTION
**Pull request by driazati**
This adds a listing of the parts of the `typing` module that are unsupported

This is also a first pass decisions on features are 'unlikely to be implemented' vs 'not implemented' so they're open to discussion

Differential Revision: [D18665628](https://our.internmc.facebook.com/intern/diff/18665628/)

### CODE REVIEW

File: docs/source/jit.rst [hunk 1]
```diff
@@ -471,6 +470,31 @@ Example (a type mismatch)
...
+
+   ":any:`typing.Any`", ":any:`typing.Any` is currently in development but not yet released"
+   ":any:`typing.NoReturn`", "Not implemented"
+   ":any:`typing.Union`", "Unlikely to be implemented"
+   ":any:`typing.Callable`", "Not implemented"
+   ":any:`typing.Literal`", "Not implemented"
+   ":any:`typing.ClassVar`", "Not implemented"
+   ":any:`typing.Final`", "This is supported for :any:`module attributes <Module Attributes>` class attribute annotations but not for functions"
+   ":any:`typing.AnyStr`", "TorchScript does not support :any:`bytes` so this type is not used"
+   "Type aliases", "Unlikely to be implemented"
```
Comment (eellison) [inline docs/source/jit.rst]: I don't think this is unlikely to be implemented. It already works when you get annotations through annotations.py.

File: docs/source/jit.rst [hunk 2]
```diff
@@ -471,6 +470,31 @@ Example (a type mismatch)
...
+   "NewType", "Unlikely to be implemented"
+   "Generics", "Unlikely to be implemented"
+
+Any other functionality from the :any:`typing` module not explitily listed in this documentation is unsupported.
```
Comment (eellison) [inline docs/source/jit.rst]: Can you add a section for https://docs.python.org/3/library/typing.html#nominal-vs-structural-subtyping ? and say that nominal typing is in development, structural typing is not.

File: docs/source/jit.rst [hunk 3]
```diff
@@ -471,6 +470,31 @@ Example (a type mismatch)
...
+   ":any:`typing.NoReturn`", "Not implemented"
+   ":any:`typing.Union`", "Unlikely to be implemented"
+   ":any:`typing.Callable`", "Not implemented"
+   ":any:`typing.Literal`", "Not implemented"
+   ":any:`typing.ClassVar`", "Not implemented"
+   ":any:`typing.Final`", "This is supported for :any:`module attributes <Module Attributes>` class attribute annotations but not for functions"
+   ":any:`typing.AnyStr`", "TorchScript does not support :any:`bytes` so this type is not used"
+   "Type aliases", "Unlikely to be implemented"
+   "NewType", "Unlikely to be implemented"
+   "Generics", "Unlikely to be implemented"
```
Comment (eellison) [inline docs/source/jit.rst]: Can you also add that https://docs.python.org/3/library/typing.html#typing.overload is in development ?

File: docs/source/jit.rst [hunk 4]
```diff
@@ -471,6 +470,31 @@ Example (a type mismatch)
...
+may be added if there is enough user demand to make it a priority.
+
+These types and features from the :mod:`typing` module are unavailble in TorchScript.
+
+.. csv-table::
+   :header: "Item", "Description"
+
+   ":any:`typing.Any`", ":any:`typing.Any` is currently in development but not yet released"
+   ":any:`typing.NoReturn`", "Not implemented"
+   ":any:`typing.Union`", "Unlikely to be implemented"
```
Comment (eellison) [inline docs/source/jit.rst]: note that we do support Optionals

**Comment by eellison** [comment]
Looks good

File: docs/source/jit.rst (hunk 4 above)
Comment (driazati) [reply docs/source/jit.rst]: I added a note, but it's also listed above in the supported types

File: docs/source/jit.rst [hunk 5]
```diff
@@ -471,6 +470,33 @@ Example (a type mismatch)
...
+   ":any:`typing.Any`", ":any:`typing.Any` is currently in development but not yet released"
+   ":any:`typing.NoReturn`", "Not implemented"
+   ":any:`typing.Union`", "Unlikely to be implemented (however :any:`typing.Optional` is supported)"
+   ":any:`typing.Callable`", "Not implemented"
+   ":any:`typing.Literal`", "Not implemented"
+   ":any:`typing.ClassVar`", "Not implemented"
+   ":any:`typing.Final`", "This is supported for :any:`module attributes <Module Attributes>` class attribute annotations but not for functions"
+   ":any:`typing.AnyStr`", "TorchScript does not support :any:`bytes` so this type is not used"
+   ":any:`typing.overload`", ":any:`typing.overload` is currently in development but not yet released"
+   "Type aliases", "Not implemented"
```
Comment (eellison) [inline docs/source/jit.rst]: Why do these four sections not have the same formatting as above ? 
":any:`typing.overload`" vs "Type aliases"

File: docs/source/jit.rst [hunk 6]
```diff
@@ -471,6 +470,33 @@ Example (a type mismatch)
      ...
 
 
+Unsupported Types
+^^^^^^^^^^^^^^^^^
+Some parts of the :mod:`typing` module are missing from TorchScript. Some of these
```
Comment (eellison) [inline docs/source/jit.rst]: nit: TorchScript does not support all of the `typing` module. Some of these....

File: docs/source/jit.rst [hunk 7]
```diff
@@ -471,6 +470,33 @@ Example (a type mismatch)
      ...
 
 
+Unsupported Types
```
Comment (eellison) [inline docs/source/jit.rst]: nit: Unsupported Typing Constructs

**Review by eellison — APPROVED** [review]
Looks good, do you mind posting a preview of what this section looks like ?

**Comment by driazati** [comment]
image

**Pull request closed without merging**